### PR TITLE
feature/cli-runner

### PR DESCRIPTION
**Pull request by brosenberg42**
Issue: https://github.com/openmpf/openmpf/issues/1003
Associated PRs:
- https://github.com/openmpf/openmpf/pull/1168
- https://github.com/openmpf/openmpf-components/pull/183
- https://github.com/openmpf/openmpf-contrib-components/pull/64
- https://github.com/openmpf/openmpf-docker/pull/118
- https://github.com/openmpf/openmpf-cpp-component-sdk/pull/90
- https://github.com/openmpf/openmpf-python-component-sdk/pull/27
- https://github.com/openmpf/openmpf.github.io/pull/120

This change is [Reviewable](https://reviewable.io/reviews/openmpf/openmpf/1168)

### CODE REVIEW

**Review by jrobble — CHANGES_REQUESTED** [review]
> Reviewed 66 of 71 files at r1, 5 of 5 files at r2, 32 of 32 files at r3.
> *[Reviewable](https://reviewable.io/reviews/openmpf/openmpf/1168)* status: all files reviewed, 7 unresolved discussions (waiting on @brosenberg42)

___
*[trunk/detection/executor/cpp/batch/Log4cxxConfig.xml, line 39 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQrWS7L9-8Qmdo86QMG:-MQrWS7L9-8Qmdo86QMH:bcramx0) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/detection/executor/cpp/batch/Log4cxxConfig.xml#L39)):*
> ```XML
>     </appender>
> 
>     <appender name="MPF-AMQ-DETECTION-CONSOLE" class="org.apache.log4j.ConsoleAppender">
> ```

TODO: Check that C++CLI Runner `docker service log` is timestamped (`Log4cxxConfig.xml`).
___
*[trunk/detection/executor/cpp/batch/MPFMessenger.h, line 76 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQrjBRO7teT4H608j2h:-MQrjBRO7teT4H608j2i:bnc781m) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/detection/executor/cpp/batch/MPFMessenger.h#L76)):*
> ```objc
> 
> /**
>  * The MPFMessenger class encapsulates ActiveMQ and Google Protocol Buffer
> ```

Please shorten this to:

```
The MPFMessenger class encapsulates ActiveMQ and Google Protocol Buffer functionality associated with the MPF component.
```

The AMQ version number is no longer accurate and we don't want to have to worry about keeping this comment up to date.
___
*[trunk/detection/executor/cpp/batch/main.cpp, line 167 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQrZgsJBZ0zWCDc1KEP:-MQrZgsJBZ0zWCDc1KEQ:b-78q4zf) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/detection/executor/cpp/batch/main.cpp#L167)):*
> ```cpp
> 
>     std::string component_name = "detection";
>     // Need to make sure COMPONENT_NAME is set because it used to determine the name of the log file.
> ```

it "is" used ...
___
*[trunk/detection/executor/cpp/streaming/StreamingExecutorLog4cxxConfig.xml, line 29 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsSLm598uS8s56vZlg:-MQsSLm66smWmC5V3Pp8:bfo1vm2) ([raw file](https://github.com/openmpf/openmpf/blob/3aa8ae131c6bba2938aa5558fd7aea9ff34ce842/trunk/detection/executor/cpp/streaming/StreamingExecutorLog4cxxConfig.xml#L29)):*
> ```XML
> 
> <log4j:configuration xmlns:log4j="http://jakarta.apache.org/log4j/">
>

**Review by jrobble — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/openmpf/openmpf/1168)* status: all files reviewed, 6 unresolved discussions (waiting on @brosenberg42)

___
*[trunk/detection/executor/cpp/streaming/StreamingExecutorLog4cxxConfig.xml, line 29 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsSLm598uS8s56vZlg:-MVccNREA-aAbhEKezC8:b-xavq5v) ([raw file](https://github.com/openmpf/openmpf/blob/3aa8ae131c6bba2938aa5558fd7aea9ff34ce842/trunk/detection/executor/cpp/streaming/StreamingExecutorLog4cxxConfig.xml#L29)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

TODO: Will nothing appear in the terminal now when running batch or streaming processes on the command line?

Components themselves now use default logging to console, executors log to file and std err?
</blockquote></details>

I tested this:

```
[mpf@openmpf-master bin]$ pwd
/home/mpf/git/openmpf-projects/openmpf/trunk/install/bin

[mpf@openmpf-master bin]$ ./amq_detection_component

[mpf@openmpf-master bin]$ cat ../share/logs/openmpf-master.mshome.net/log/detection.log
...
2021-03-12 17:49:26,539 ERROR [0x7ff86ca328c0] org.mitre.mpf.detection:95 - ERROR: the activemq broker uri, component library name, and request queue name must be supplied as arguments
```

In the past I used to run `amq_detection_component` manually for debugging purposes, sometimes on a production machine, but honestly, I haven't done it in a long time.

What inspired you to remove the console output?

**Review by brosenberg42 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/openmpf/openmpf/1168)* status: all files reviewed, 6 unresolved discussions (waiting on @brosenberg42 and @jrobble)

___
*[trunk/detection/executor/cpp/streaming/StreamingExecutorLog4cxxConfig.xml, line 29 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsSLm598uS8s56vZlg:-MVpm7Rz8d9qhdNgtC2X:b-rjp5jv) ([raw file](https://github.com/openmpf/openmpf/blob/3aa8ae131c6bba2938aa5558fd7aea9ff34ce842/trunk/detection/executor/cpp/streaming/StreamingExecutorLog4cxxConfig.xml#L29)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

I tested this:

```
[mpf@openmpf-master bin]$ pwd
/home/mpf/git/openmpf-projects/openmpf/trunk/install/bin

[mpf@openmpf-master bin]$ ./amq_detection_component

[mpf@openmpf-master bin]$ cat ../share/logs/openmpf-master.mshome.net/log/detection.log
...
2021-03-12 17:49:26,539 ERROR [0x7ff86ca328c0] org.mitre.mpf.detection:95 - ERROR: the activemq broker uri, component library name, and request queue name must be supplied as arguments
```

In the past I used to run `amq_detection_component` manually for debugging purposes, sometimes on a production machine, but honestly, I haven't done it in a long time.

What inspired you to remove the console output?
</blockquote></details>

Your comments don't make any sense. 
* The only change I made in this file was changing the name of the log file.  I removed `appxConsoleAppender` because wasn't attached to any logger so it wasn't doing anything.
* You are commenting on the configuration for streaming component executor, but starting the batch executor.
* When I perform the exact same test as above, I see the same results on develop. However, when I switched to using cli runner branch, I see the error message in the console.
___
*[trunk/interop/src/main/java/org/mitre/mpf/interop/JsonTrackOutputObject.java, line 151 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsOaEk738Hmzdl8HQo:-MVppir726OD1luiY0Lc:b-68nlfi) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/interop/src/main/java/org/mitre/mpf/interop/JsonTrackOutputObject.java#L151)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

When will the source and id be null? Tests?
</blockquote></details>

cli runner output. It doesn't run pipelines, so certain fields don't make sense.
___
*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/BaseServiceLauncher.java, line 140 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsKeMT78DJcKUyYVld:-MVpqFSe4A9q1rgsyizI:bmrijav) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/node-manager/src/main/java/org/mitre/mpf/nms/BaseServiceLauncher.java#L140)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

I agree that logging warnings should not depend on logging debug.

What should we do if `LOG.isErrorEnabled()` but not `LOG.isWarnEnabled()`?

---

Looking at the `node-manager` log in a local dev. env. deployment:

```
2021-01-15 10:27:02,678 WARN [Thread-101] o.m.m.n.BaseServiceLauncher - Node stderr [PythonTestComponent]: 2021-01-15 10:27:02,668 INFO  [test_component.py:39] - Creating instance of TestComponent
2021-01-15 10:27:04,525 WARN [Thread-101] o.m.m.n.BaseServiceLauncher - Node stderr [PythonTestComponent]: 2021-01-15 10:27:04,525 INFO  [org.mitre.mpf.detection:0] - Creating ActiveMQ consumer for queue: MPF.DETECTION_PYTHONTEST_REQUEST
2021-01-15 10:27:04,538 WARN [Thread-101] o.m.m.n.BaseServiceLauncher - Node stderr [PythonTestComponent]: 2021-01-15 10:27:04,538 INFO  [org.mitre.mpf.detection:0] - Completed initialization of openmpf-master.mshome.net:PythonTestComponent:1.
2021-01-15 10:27:09,202 WARN [Thread-90] o.m.m.n.BaseServiceLauncher - Node stderr [OcvDnnDetection]: 2021-01-15 10:27:09,202 INFO [0x7fb16a1128c0] org.mitre.mpf.detection:94 - Creating ActiveMQ consumer for queue: MPF.DETECTION_DNNCV_REQUEST
2021-01-15 10:27:09,211 WARN [Thread-90] o.m.m.n.BaseServiceLauncher - Node stderr [OcvDnnDetection]: 2021-01-15 10:27:09,211 INFO [0x7fb16a1128c0] OcvDnnDetection:69 - Initializing OcvDnn
2021-01-15 10:27:09,211 WARN [Thread-90] o.m.m.n.BaseServiceLauncher - Node stderr [OcvDnnDetection]: 2021-01-15 10:27:09,211 INFO [0x7fb16a1128c0] org.mitre.mpf.detection:94 - Completed initialization of openmpf-master.mshome.net:OcvDnnDetection:1.
```

These lines seem like they should be at the `INFO` level, not `WARN`.
</blockquote></details>

I don't understand. I can't change the log level to both `ERROR` and `INFO`. The change was made for debugging, it should just be removed.

**Review by jrobble — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/openmpf/openmpf/1168)* status: all files reviewed, 4 unresolved discussions (waiting on @brosenberg42)

___
*[trunk/detection/executor/cpp/streaming/StreamingExecutorLog4cxxConfig.xml, line 29 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsSLm598uS8s56vZlg:-MVrGeTRAsoFlawQkLs6:bf1opjv) ([raw file](https://github.com/openmpf/openmpf/blob/3aa8ae131c6bba2938aa5558fd7aea9ff34ce842/trunk/detection/executor/cpp/streaming/StreamingExecutorLog4cxxConfig.xml#L29)):*
<details><summary><i>Previously, brosenberg42 wrote…</i></summary><blockquote>

Your comments don't make any sense. 
* The only change I made in this file was changing the name of the log file.  I removed `appxConsoleAppender` because wasn't attached to any logger so it wasn't doing anything.
* You are commenting on the configuration for streaming component executor, but starting the batch executor.
* When I perform the exact same test as above, I see the same results on develop. However, when I switched to using cli runner branch, I see the error message in the console.
</blockquote></details>

Sorry for the confusion. I saw that you removed `appxConsoleAppender` from the batch logger config as well. I tested batch since that's what I use more often.

I must be misremembering. I thought that the old executor code output to the console, but I was probably just tailing the log. When reviewing this PR I saw `appxConsoleAppender` and assumed it used to work.

I wrote my last comment on March 12th, but the binary on my dev. VM is from March 10th, so yeah, it's old. I must have only built the cli-runner code in Docker. I just ran the test again after building and it behaves as you described, which is an improvement.
___
*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/BaseServiceLauncher.java, line 140 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsKeMT78DJcKUyYVld:-MVrMwHi-j1zedIx7DuU:b75400i) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/node-manager/src/main/java/org/mitre/mpf/nms/BaseServiceLauncher.java#L140)):*
> I agree that logging warnings should not depend on logging debug.

The previous code would only ever log warnings if debug was enabled. Your new code sets up `errReceiver` based on if warnings are enabled, which is a separate check. At first glance I thought that was a good idea.

Looking into it more though, a lot of the content on the stderr stream is not related to errors, so I would only set it up if debug is enabled, as we did before. 

I agree that the content may be a good candidate for `INFO` instead of `DEBUG`, but we've been living without it at the `INFO` level, so there's no reason to enable it now, unless you think I has value.

The node-manager log will really only be used on a dev. env. anyway, so this won't affect end users, unless we end up back-porting this change along with others to one of the users' older Docker deployments.

**Review by brosenberg42 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/openmpf/openmpf/1168)* status: 67 of 71 files reviewed, 4 unresolved discussions (waiting on @jrobble)

___
*[trunk/detection/executor/cpp/batch/MPFMessenger.h, line 76 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQrjBRO7teT4H608j2h:-MWdk2m62_ipCl-9jF24:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/detection/executor/cpp/batch/MPFMessenger.h#L76)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

Please shorten this to:

```
The MPFMessenger class encapsulates ActiveMQ and Google Protocol Buffer functionality associated with the MPF component.
```

The AMQ version number is no longer accurate and we don't want to have to worry about keeping this comment up to date.
</blockquote></details>

Done.
___
*[trunk/detection/executor/cpp/batch/main.cpp, line 167 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQrZgsJBZ0zWCDc1KEP:-MWdmkvw0Q1dwmiE7l5A:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/detection/executor/cpp/batch/main.cpp#L167)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

it "is" used ...
</blockquote></details>

Done.
___
*[trunk/mpf-rest-api/src/main/java/org/mitre/mpf/rest/api/node/DeployedServiceModel.java, line 43 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsMSysAX_IUk2ThaY2:-MWdnr-w5Rw8k0Zypg8v:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/mpf-rest-api/src/main/java/org/mitre/mpf/rest/api/node/DeployedServiceModel.java#L43)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

Please remove this inline comment.
</blockquote></details>

Done.
___
*[trunk/node-manager/src/main/java/org/mitre/mpf/nms/BaseServiceLauncher.java, line 140 at r2](https://reviewable.io/reviews/openmpf/openmpf/1168#-MQsKeMT78DJcKUyYVld:-MWdndQB8HKPHIj7xzD4:b-896fix) ([raw file](https://github.com/openmpf/openmpf/blob/9941c0a1b88e43a6d67a9ae9458e116e2984def7/trunk/node-manager/src/main/java/org/mitre/mpf/nms/BaseServiceLauncher.java#L140)):*
<details><summary><i>Previously, jrobble (Jeff Robble) wrote…</i></summary><blockquote>

> I agree that logging warnings should not depend on logging debug.

The previous code would only ever log warnings if debug was enabled. Your new code sets up `errReceiver` based on if warnings are enabled, which is a separate check. At first glance I thought that was a good idea.

Looking into it more though, a lot of the content on the stderr stream is not related to errors, so I would only set it up if debug is enabled, as we did before. 

I agree that the content may be a good candidate for `INFO` instead of `DEBUG`, but we've been living without it at the `INFO` level, so there's no reason to enable it now, unless you think I has value.

The node-manager log will really only be used on a dev. env. anyway, so this won't affect end users, unless we end up back-porting this change along with others to one of the users' older Docker deployments.
</blockquote></details>

Done.

**Review by jrobble — APPROVED** [review]
> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/openmpf/openmpf/1168)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @brosenberg42)
